File: frontend/src/api/copilot.ts
import { fetchApi } from './client';

export interface NoteExplanationResponse {
  note_id: string;
  markdown: string;
  referenced_headings: string[];
}

export interface NoteSummaryResponse {
  note_id: string;
  markdown: string;
  key_points: string[];
}

export interface LinkSuggestion {
  target_note_path: string;
  target_title: string;
  reason: string;
}

export interface SuggestLinksResponse {
  note_id: string;
  suggestions: LinkSuggestion[];
}

export interface TagSuggestion {
  tag: string;
  confidence: number;
  reason: string;
}

export interface SuggestTagsResponse {
  note_id: string;
  suggestions: TagSuggestion[];
}

export interface StructureIssue {
  type: 'missing_heading' | 'deeply_nested' | 'long_paragraph' | 'logical_gap';
  location: string;
  issue: string;
  suggestion: string;
}

export interface SuggestStructureResponse {
  note_id: string;
  suggestions: StructureIssue[];
}

export interface ProposePatchResponse {
  job_id: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
}

export interface ProposePatchResult {
  note_id: string;
  proposal_id: string;
  diff: string;
}

export interface ChatMessage {
  role: 'user' | 'assistant';
  content: string;
}

export interface ChatResponse {
  note_id: string;
  message: ChatMessage;
}

export const copilotApi = {
  explain: (noteId: string) =>
    fetchApi<NoteExplanationResponse>(`/copilot/explain/${noteId}`, { method: 'POST' }),

  summarize: (noteId: string) =>
    fetchApi<NoteSummaryResponse>(`/copilot/summarize/${noteId}`, { method: 'POST' }),

  suggestLinks: (noteId: string) =>
    fetchApi<SuggestLinksResponse>(`/copilot/suggest-links/${noteId}`, { method: 'POST' }),

  suggestTags: (noteId: string) =>
    fetchApi<SuggestTagsResponse>(`/copilot/suggest-tags/${noteId}`, { method: 'POST' }),

  suggestStructure: (noteId: string) =>
    fetchApi<SuggestStructureResponse>(`/copilot/suggest-structure/${noteId}`, { method: 'POST' }),

  proposePatch: (noteId: string, instruction: string) =>
    fetchApi<ProposePatchResponse>(`/copilot/propose-patch/${noteId}`, {
      method: 'POST',
      body: JSON.stringify({ instruction }),
    }),

  chat: (noteId: string, message: string, history: ChatMessage[]) =>
    fetchApi<ChatResponse>(`/copilot/chat/${noteId}`, {
      method: 'POST',
      body: JSON.stringify({ message, history }),
    }),
};

File: src/schemas/copilot.py
"""Copilot schemas — F11-01 through F11-11."""
from pydantic import BaseModel, Field
from uuid import UUID
from typing import Literal

# Base
class CopilotAction(BaseModel):
    note_id: UUID
    instruction: str | None = None

# Explain / Summarize
class NoteExplanationResponse(BaseModel):
    note_id: UUID
    markdown: str = Field(description="Markdown explanation")
    referenced_headings: list[str] = Field(default_factory=list)

class NoteSummaryResponse(BaseModel):
    note_id: UUID
    markdown: str = Field(description="100-300 word summary")
    key_points: list[str] = Field(default_factory=list)

# Suggest Links
class LinkSuggestion(BaseModel):
    target_note_path: str
    target_title: str
    reason: str

class SuggestLinksResponse(BaseModel):
    note_id: UUID
    suggestions: list[LinkSuggestion]

# Suggest Tags
class TagSuggestion(BaseModel):
    tag: str
    confidence: float = Field(ge=0.0, le=1.0)
    reason: str

class SuggestTagsResponse(BaseModel):
    note_id: UUID
    suggestions: list[TagSuggestion]

# Suggest Structure
class StructureIssue(BaseModel):
    type: Literal["missing_heading", "deeply_nested", "long_paragraph", "logical_gap"]
    location: str = Field(description="Heading path or line range")
    issue: str
    suggestion: str

class SuggestStructureResponse(BaseModel):
    note_id: UUID
    suggestions: list[StructureIssue]

# Propose Patch
class ProposePatchRequest(BaseModel):
    instruction: str = Field(min_length=1, max_length=1000)

class ProposePatchResponse(BaseModel):
    note_id: UUID
    proposal_id: UUID
    diff: str = Field(description="Unified diff")


class ProposePatchEnqueueResponse(BaseModel):
    """Response when a propose_patch job is enqueued."""
    job_id: UUID
    status: str = "pending"

# Chat
class ChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str

class ChatRequest(BaseModel):
    note_id: UUID
    message: str
    history: list[ChatMessage] = Field(default_factory=list)

class ChatResponse(BaseModel):
    note_id: UUID
    message: ChatMessage
